docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,1963 @@
+{
```

Comment:
<pre class="ql-syntax" spellcheck="false">Line #88.            min_distances = min_dist * (1 - np.arange(max_attempts[i]) * decay[i])
</pre><p>I find this line hard to comprehend after stepping away from it. I think it may be more obvious if we use a <code>numpy</code> built-in function here:</p><pre class="ql-syntax" spellcheck="false">min_distances = np.linspace(min_dist, 0, num=int(1.0 / decay[i]), endpoint=False) 
</pre><p>Then, we can get rid of <code>max_attempts</code> altogether. <code>np.linspace</code> is pretty common when you want to grab evenly spaces points along a line that starts/stops between two specific points. I think this is way more clear.</p><p> </p><p>And then, since <code>np.linspace</code> has boundary guarantees,  I wonder if we even need to perform <code>np.clip</code>??</p><p> </p><p>As you can see, we continue to find ways to make the code more readable AND we are doing so by removing lines of code (instead of adding). </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

Reply:
> As you can see, we continue to find ways to make the code more readable AND we are doing so by removing lines of code (instead of adding).

Correct! Thanks for the suggestion!! 

I think `endpoint=False` can cause a problem. So, by `endpoint=False`, we are excluding 0 in `min_distances`. So, let's say the last element in `min_distnaces` (i.e. smallest value) is `0.1`. But, what if the discord has distance of 0.09 to its nearest neighbor? (btw, this is one of those cases where we might end up with a lot of candidates in the first phase as most sequences are very close to each other). 

We may add a new input `min_discord_distance` that can be provided by user (e.g. default value can be 0.001). I know `min_dist = 0` does not make sense for discord and that is why you did  `endpoint=False`. The challenge is that we  do not know where to stop in `np.linspace(start, stop=???, ...)`.  `min_dist=0` does not make sense but what is the best next value that we should consider for `stop` in `np.linspace(...)`.

Reply:
<p>Instead of adding yet another parameter of <code>min_discord_distance</code>, I would add a new global parameter to <code>stumpy/config.py</code>  that is <code>STUMPY_MIN_DISCORD_DISTANCE = 1e-6</code>  and then  do:</p><p> </p><pre class="ql-syntax" spellcheck="false">min_distances = np.linspace(min_dist, config.STUMPY_MIN_DISCORD_DISTANCE, num=int(1.0 / decay[i]), endpoint=True)
</pre><p> </p><p>I don't know what the minimum distance is but, in my experience, <code>1e-6</code> is essentially zero due to numerical precision and I really don't see users hoping for discords that are smaller than <code>1e-6</code>.  Because it really makes very little sense. Even <code>1e-6</code> is questionable as a lower bound. If we need to reduce <code>min_dist</code> to such a small number and we still can't find a discord then the probability of finding a discord is just too low. Ultimately, if the user REALLY wants to then they can change <code>config.STUMP_MIN_DISCORD_DISTANCE</code> to some other smaller value but that is something that they really need to be consciously thinking through. </p><p> </p><p>What do you think? Personally, I'm not interested in solving ALL possible edge cases and this feels like an extreme edge case.  </p>

Reply:
<p>Since <code>min_dist</code> is an input parameter, that "should" be the absolute minimum distance allowed. However, when <code>min_dist = None</code> then we should make the absolute minimum distance be <code>config.STUMPY_MIN_DISCORD_DISTANCE</code>    </p>

Reply:
> in my experience, `1e-6` is essentially zero due to numerical precision and I really don't see users hoping for discords that are smaller than `1e-6`. Because it really makes very little sense. Even `1e-6` is questionable as a lower bound. If we need to reduce `min_dist` to such a small number and we still can't find a discord then the probability of finding a discord is just too low.

Exactly! It is important to note that the array `min_distances = np.linspace(...)` depends on `min_dist` and `decay[i]`. So, if `min_dist=10`  and `decay[i]=0.2`, then, with `endpoint=False`, our lowest value in array `min_distances` would be 2, which might be still reasonable. (I mean... it is not close to 1e-6)

So, I agree with you on not solving ALL possible edge cases. But, since we do not know the value of `min_dist` (and `decay`) in advance, the lowest value in array `min_distances` when `endpoint=False` might not be close to 1e-6, and thus, maybe (?), not an extreme case. (Maybe I am thinking too much about these possible scenarios 😄 )

---

Your suggestion on creating global variable is cleaner and better. Thanks! So, just to confirm, I will go with:
```
min_distances = np.linspace(min_dist, config.STUMPY_MIN_DISCORD_DISTANCE, num=int(1.0 / decay[i]), endpoint=True)
```

Should I use `1e-6` for now instead of `config.STUMPY_MIN_DISCORD_DISTANCE`?

Reply:
> Since `min_dist` is an input parameter, that "should" be the absolute minimum distance allowed. However, when `min_dist = None` then we should make the absolute minimum distance be `config.STUMPY_MIN_DISCORD_DISTANCE`

Are you talking about line # 59 (in function `_discords`)? I think there is a misunderstanding here and that shows my code was not clear! So, sorry about that. I should have been more careful in designing this. 

By `None`, I meant the user do not know a good value for `min_dist`. So, we set it to highest possible value (let's call it `max_dist` which is `2*\sqrt(m)` in z-normalize case). Also, if `min_dist` is set to a number higher than `max_dist`, it should be replaced by `max_dist`).

------

So, to wrap up... I can just use default `np.inf` instead of `None` for input `min_dist` . Then, I can rewrite line # 59, 60 as follows:
```
min_dist = min(min_dist, 2 * np.sqrt(m))
```

Notes:
* why not setting default to `2 * np.sqrt(m)`?  
User does not need to think why  `2 * np.sqrt(m)`, and I thought default `min_dist = np.inf` might sound more digestible (?!?)

Reply:
<pre class="ql-syntax" spellcheck="false"># config.STUMPY_MIN_DISCORD_DISTANCE = 1e-6
min_distances = np.linspace(min_dist, 1e-6, num=int(1.0 / decay[i]), endpoint=True)
</pre><p>Maybe just write it like this so we remember to move this to <code>config.py</code> later. And, yes, <code>endpoint=True</code> is fine  </p>

Reply:
<blockquote>&nbsp;I can just use default&nbsp;<code> 2 * np.sqrt(m)</code>&nbsp;instead of&nbsp;<code>None</code>&nbsp;for input&nbsp;<code>min_dist</code></blockquote><p>I don't think None must be the default as I don't think you will be able to set <code>min_dist=2.0*np.sqrt(m)</code> in the function definition. Let's not change Lines 59-60 and keep it the same. To be honest, I just forgot that you had set <code>min_dist</code>  to <code>2.0*np.dqrt(m)</code>  and it speaks to the problem that we are changing <code>min_dist</code> in our code. I think I mentioned this before but I'm not convinced that we need to redesign the logic inside the for-loop. It's just too hard to follow.  I tried really quickly to change it but everything is all tangled together. I think a different design is preferred. Will you spend some time focusing on this please? Try writing it from scratch again and see what alternative you can come up with (even if it is slow).</p>

Reply:
> as I don't think you will be able to set `min_dist=2.0*np.sqrt(m)` in the function definition

Yep! I missed that part when I was thinking!! I will try to write the function `_discords` again.

Reply:
<p>Given the line:</p><pre class="ql-syntax" spellcheck="false">if np.all(include == False) or min_dist &lt;= 0:
</pre><p>I think that either we just allow <code>min_dist</code> to be as small as zero or we should check if <code>min_dist &lt;= config.STUMPY_MIN_DISCORD_DISTANCE</code> </p><p> </p><p>Additionally, I may need to concede and, instead of an outer <code>for</code> loop, it may be less constraining to make it a <code>while</code> loop. Like:</p><p> </p><pre class="ql-syntax" spellcheck="false">discords_idx = []
while len(discords_idx) &lt; k:
    # Add other <code>break</code> condition
    # keep looking for a discord
    IDX, P, I = _find_discord(T, m, M_T, Σ_T, is_cands, approx_P)
    discords_idx.append(IDX)
    discords_dist.append(P)
    discords_nn_idx.append(I)
    
</pre>

Reply:
<p>I do hate <code>while</code> loops but it may be more natural in this case </p>

Reply:
> I think that either we just allow min_dist to be as small as zero or we should check if min_dist <= config.STUMPY_MIN_DISCORD_DISTANCE

I will change the if-condition to: 
```
if np.all(include == False) or min_dist <= config.STUMPY_MIN_DISCORD_DISTANCE:
```


> I may need to concede and, instead of an outer `for` loop, it may be less constraining to make it a `while` loop

I was thinking about the same thing today. Because, I thought, if function `_discords` returns more than one discord (from a set of candidates), then `while` would be cleaner... but, I know you hate it (as I noticed before that you've opened an issue to replace `while` loop  in one of STUMPY modules.) so I didn't mention it :D 

---

Btw, do you prefer variable names `IDX, P, I`? (or you just wrote them here for the sake of simplicity)?

Reply:
<p>I just wrote it for simplicity</p>

Reply:
Yes... It is good!